File: Demo/Application/Features/Amex/BraintreeDemoAmexViewController.m
#import "BraintreeDemoAmexViewController.h"
@import BraintreeAmericanExpress;
@import BraintreeCard;

@interface BraintreeDemoAmexViewController ()

@property (nonatomic, strong) BTAPIClient *apiClient;
@property (nonatomic, strong) BTCardClient *cardClient;
@property (nonatomic, strong) BTAmericanExpressClient *amexClient;

@end

@implementation BraintreeDemoAmexViewController

- (instancetype)initWithAuthorization:(NSString *)authorization {
    self = [super initWithAuthorization:authorization];
    if (self) {
        _apiClient = [[BTAPIClient alloc] initWithAuthorization:authorization];
        _amexClient = [[BTAmericanExpressClient alloc] initWithAPIClient:_apiClient];
        _cardClient = [[BTCardClient alloc] initWithAPIClient:_apiClient];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = NSLocalizedString(@"Amex", nil);

    UIButton *validCardButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [validCardButton setTitle:NSLocalizedString(@"Valid card", nil) forState:UIControlStateNormal];
    [validCardButton addTarget:self action:@selector(tappedValidCard) forControlEvents:UIControlEventTouchUpInside];

    UIButton *insufficientPointsCardButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [insufficientPointsCardButton setTitle:NSLocalizedString(@"Insufficient points card", nil) forState:UIControlStateNormal];
    [insufficientPointsCardButton addTarget:self action:@selector(tappedInsufficientPointsCard) forControlEvents:UIControlEventTouchUpInside];

    UIButton *ineligibleCardButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [ineligibleCardButton setTitle:NSLocalizedString(@"Ineligible card", nil) forState:UIControlStateNormal];
    [ineligibleCardButton addTarget:self action:@selector(tappedIneligibleCard) forControlEvents:UIControlEventTouchUpInside];

    UIStackView *stackView = [[UIStackView alloc] initWithArrangedSubviews:@[validCardButton, insufficientPointsCardButton, ineligibleCardButton]];
    stackView.axis = UILayoutConstraintAxisVertical;
    stackView.distribution = UIStackViewDistributionEqualSpacing;
    stackView.spacing = 20;
    stackView.translatesAutoresizingMaskIntoConstraints = NO;

    [self.view addSubview:stackView];
    [NSLayoutConstraint activateConstraints:@[
        [stackView.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor],
        [stackView.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor]
    ]];
}

- (void)tappedValidCard {
    [self getRewardsForCardNumber:@"371260714673002"];
}

- (void)tappedInsufficientPointsCard {
    [self getRewardsForCardNumber:@"371544868764018"];
}

- (void)tappedIneligibleCard {
    [self getRewardsForCardNumber:@"378267515471109"];
}

- (void)getRewardsForCardNumber:(NSString *)cardNumber {
    BTCard *card = [BTCard new];
    card.number = cardNumber;
    card.expirationMonth = @"12";
    card.expirationYear = @"2025";
    card.cvv = @"1234";
    
    self.progressBlock(@"Tokenizing Card");

    [self.cardClient tokenizeCard:card completion:^(BTCardNonce *tokenized, NSError *error) {
        if (error) {
            self.progressBlock(error.localizedDescription);
            NSLog(@"Error: %@", error);
            return;
        }

        self.progressBlock(@"Amex - getting rewards balance");
        [self.amexClient getRewardsBalanceForNonce:tokenized.nonce currencyIsoCode:@"USD" completion:^(BTAmericanExpressRewardsBalance *rewardsBalance, NSError *error) {
            if (error) {
                self.progressBlock(error.localizedDescription);
            } else if (rewardsBalance.errorCode) {
                self.progressBlock([NSString stringWithFormat:@"%@: %@", rewardsBalance.errorCode, rewardsBalance.errorMessage]);
            } else {
                self.progressBlock([NSString stringWithFormat:@"%@ %@, %@ %@", rewardsBalance.rewardsAmount, rewardsBalance.rewardsUnit, rewardsBalance.currencyAmount, rewardsBalance.currencyIsoCode]);
            }
        }];
    }];
}

@end

File: Demo/Application/Features/Amex/BraintreeDemoAmexViewController.swift
import Foundation
import BraintreeAmericanExpress
import BraintreeCard

class BraintreeDemoAmexViewController: BraintreeDemoPaymentButtonBaseViewController {

    lazy var amexClient = BTAmericanExpressClient(apiClient: apiClient)
    lazy var cardClient = BTCardClient(apiClient: apiClient)

    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Amex"
    }

    override func createPaymentButton() -> UIView! {
        let validCardButton = createButton(title: "Valid card", action: #selector(tappedValidCard))
        let insufficientPointsCardButton = createButton(title: "Insufficient points card", action: #selector(tappedInsufficientPointsCard))
        let ineligibleCardButton = createButton(title: "Ineligible card", action: #selector(tappedIneligibleCard))

        let stackView = UIStackView(arrangedSubviews: [validCardButton, insufficientPointsCardButton, ineligibleCardButton])
        stackView.axis = .vertical
        stackView.alignment = .center
        stackView.distribution = .fillEqually
        stackView.translatesAutoresizingMaskIntoConstraints = false

        return stackView
    }

    @objc func tappedValidCard() {
        getRewards(for: "371260714673002")
    }

    @objc func tappedInsufficientPointsCard() {
        getRewards(for: "371544868764018")
    }

    @objc func tappedIneligibleCard() {
        getRewards(for: "378267515471109")
    }

    private func getRewards(for cardNumber: String) {
        let card = BTCard()
        card.number = cardNumber
        card.expirationMonth = "12"
        card.expirationYear = generateFutureYear()
        card.cvv = "1234"

        progressBlock("Tokenizing Card")

        cardClient.tokenize(card) { tokenizedCard, error in
            guard let tokenizedCard else {
                self.progressBlock(error?.localizedDescription)
                return
            }

            self.progressBlock("Amex - getting rewards balance")

            self.amexClient.getRewardsBalance(forNonce: tokenizedCard.nonce, currencyISOCode: "USD") { rewardsBalance, error in
                guard let rewardsBalance else {
                    self.progressBlock(error?.localizedDescription)
                    return
                }

                if let errorCode = rewardsBalance.errorCode, let errorMessage = rewardsBalance.errorMessage  {
                    self.progressBlock("\(errorCode): \(errorMessage)")
                } else if let rewardsAmount = rewardsBalance.rewardsAmount,
                          let rewardsUnit = rewardsBalance.rewardsUnit,
                          let currencyAmount = rewardsBalance.currencyAmount,
                          let currencyIsoCode = rewardsBalance.currencyIsoCode {
                    self.progressBlock("\(rewardsAmount) \(rewardsUnit), \(currencyAmount) \(currencyIsoCode)")
                }
            }
        }
    }

    private func generateFutureYear() -> String {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yy"

        let futureYear = Calendar.current.date(byAdding: .year, value: 3, to: Date())!
        return dateFormatter.string(from: futureYear)
    }

    // TODO: move this helper into BraintreeDemoPaymentButtonBaseViewController once converted so all buttons share the same characteristics
    private func createButton(title: String, action: Selector) -> UIButton {
        let button = UIButton(type: .system)
        button.setTitle(title, for: .normal)
        button.setTitleColor(.blue, for: .normal)
        button.setTitleColor(.lightGray, for: .highlighted)
        button.setTitleColor(.lightGray, for: .disabled)
        button.translatesAutoresizingMaskIntoConstraints = false
        button.addTarget(self, action: action, for: .touchUpInside)
        return button
    }
}
